app/services/experiment.py

```diff
@@ -0,0 +1,269 @@
+                    exc_info=e,
+                )
+
+        count = self.experiment_run_queue.qsize()
+        if count > 0:
+            self.logger.info(
+                "Workflow queue has been initialized with "
+                + f"{count} workflows to execute"
+            )
+
+    async def init_image_build_queue(self) -> None:
+        run_ids = (
+            await ExperimentTemplate.find(
+                ExperimentTemplate.state != TemplateState.FINISHED,
+                ExperimentTemplate.state != TemplateState.CRASHED,
+            )
+            .sort(+ExperimentTemplate.updated_at)
+            .project(ExperimentTemplateId)
+            .to_list()
+        )
+        for run_id in run_ids:
+            await self.add_image_to_build(run_id.id)
+
+        count = self.image_building_queue.qsize()
+        if count > 0:
+            self.logger.info(
+                "Docker image queue has been initialized with "
+                + f"{count} images to build and push"
+            )
+
+    def close_docker_connection(self) -> None:
+        if self.docker_client:
+            self.docker_client.close()
+
+    # TODO
+    def download_files(self) -> None:
+        pass
+
+    # TODO
+    def list_files(self) -> None:
+        pass
+
+    async def add_run_to_execute(self, er_id: PydanticObjectId) -> None:
+        await self.experiment_run_queue.put(er_id)
+
+    async def get_run_to_execute(self) -> PydanticObjectId:
+        return await self.experiment_run_queue.get()
+
+    async def add_image_to_build(self, temp_id: PydanticObjectId) -> None:
+        await self.image_building_queue.put(temp_id)
+
+    async def get_image_to_build(self) -> PydanticObjectId:
+        return await self.image_building_queue.get()
+
+    async def schedule_experiment_runs(self) -> None:
+        semaphore = asyncio.Semaphore(settings.MAX_PARALLEL_CONTAINERS)
+        while True:
+            er_id = await self.get_run_to_execute()
+            asyncio.create_task(self.execute_experiment_run(er_id, semaphore))
+
+    async def schedule_image_building(self) -> None:
+        semaphore = asyncio.Semaphore(settings.MAX_PARALLEL_CONTAINERS)
+        while True:
+            temp_id = await self.get_image_to_build()
+            asyncio.create_task(self.build_and_push_image(temp_id, semaphore))
+
+    async def execute_experiment_run(
+        self, exp_run_id: PydanticObjectId, semaphore: asyncio.Semaphore
+    ) -> None:
+        async with semaphore:
+            experiment_run, experiment = await self.init_run(exp_run_id)
+
+            error_msg = await ReanaService.run_workflow(experiment_run, experiment)
+            await ReanaService.postprocess_workflow(experiment_run, error_msg)
+
+            if not error_msg:
+                experiment_run.update_state(RunState.FINISHED)
+                await experiment_run.replace()
+            elif experiment_run.run_number < settings.MAX_EXPERIMENT_RUN_ATTEMPTS - 1:
```

Comment:
I will quickly change the logic of the `run_number` attribute as I've mentioned it to you previously

Reply:
What I've done is that I renamed the `run_number` to `retry_number` and this attribute is being incremented only when a run crashes. If we were to manually execute another experiment run later, regardless on the fact whether previous runs have or have not crashed, the newly created run has `retry_number` of 0